server/tests/test_sample_all_ballots.py

```diff
@@ -382,3 +380,186 @@ def test_all_ballots_audit(
+            "'' is too short",
+        ),
+        (
+            [
+                {
+                    "batchName": None,
+                    "batchType": "Provisional",
+                    "choiceResults": {
+                        choice["id"]: choice["numVotes"] / 4
+                        for choice in contest["choices"]
+                    },
+                }
+            ],
+            "None is not of type 'string'",
+        ),
+        (
+            [
+                {
+                    "batchType": "Provisional",
+                    "choiceResults": {
+                        choice["id"]: choice["numVotes"] / 4
+                        for choice in contest["choices"]
+                    },
+                }
+            ],
+            "'batchName' is a required property",
+        ),
+        (
+            [
+                {
+                    "batchName": "Batch 1",
+                    "batchType": "bad type",
+                    "choiceResults": {
+                        choice["id"]: choice["numVotes"] / 4
+                        for choice in contest["choices"]
+                    },
+                }
+            ],
+            "'bad type' is not one of ['Absentee By Mail', 'Advance', 'Election Day', 'Provisional', 'Other']",
+        ),
+        (
+            [
+                {
+                    "batchName": "Batch 1",
+                    "batchType": "bad type",
+                    "choiceResults": {
+                        choice["id"]: choice["numVotes"] / 4
+                        for choice in contest["choices"]
+                    },
+                }
+            ],
+            "'bad type' is not one of ['Absentee By Mail', 'Advance', 'Election Day', 'Provisional', 'Other']",
+        ),
+        (
+            [
+                {
+                    "batchName": "Batch 1",
+                    "batchType": "Provisional",
+                    "choiceResults": {
+                        choice["id"]: choice["numVotes"] / 4
+                        for choice in contest["choices"][:1]
+                    },
+                }
+            ],
+            "Invalid choice ids for batch Batch 1",
+        ),
+        (
+            [
+                {
+                    "batchName": "Batch 1",
+                    "batchType": "Provisional",
+                    "choiceResults": {"not a real id": 0},
+                }
+            ],
+            "Invalid choice ids for batch Batch 1",
+        ),
+        (
+            [
+                {
+                    "batchName": "Batch 1",
+                    "batchType": "Provisional",
+                    "choiceResults": {
+                        choice["id"]: choice["numVotes"]
+                        for choice in contest["choices"]
+                    },
+                }
+            ],
+            "Total results (2000000) cannot be greater than the number of ballots reported in the ballot manifest (1000000)",
```

Comment:
I don't think we want to enforce this. If the manifest undercounted ballots, the audit should be able to deal with that.

Reply:
fixed